packages/graphiql/src/state/actions.ts

```diff
@@ -0,0 +1,54 @@
+import { EditorContexts } from './types';
+
+export enum ActionTypes {
+  OperationRequested = 'OperationRequested',
+  EditorLoaded = 'EditorLoaded',
+  OperationChanged = 'OperationChanged',
+  VariablesChanged = 'VariablesChanged',
+  OperationSucceeded = 'OperationSucceeded',
+  OperationErrored = 'OperationErrored',
+}
+
+export type ErrorPayload = { error: Error; sessionId: number };
+export type SuccessPayload = { sessionId: number; result: string };
+export type ChangeValuePayload = { sessionId: number; value: string };
+
+export type Action =
+  | OperationRequestedAction
+  | EditorLoadedAction
+  | OperationChangedAction
+  | VariablesChangedAction
+  | OperationSucceededAction
+  | OperationErroredAction;
+
+export type OperationRequestedAction = {
+  type: ActionTypes.OperationRequested;
+};
+
+export type EditorLoadedAction = {
+  type: ActionTypes.EditorLoaded;
+  payload: {
+    context: EditorContexts;
+    editor: CodeMirror.Editor;
+  };
+};
+
+export type OperationChangedAction = {
+  type: ActionTypes.OperationChanged;
+  payload: ChangeValuePayload;
+};
+
+export type VariablesChangedAction = {
+  type: ActionTypes.VariablesChanged;
+  payload: ChangeValuePayload;
+};
+
+export type OperationSucceededAction = {
+  type: ActionTypes.OperationSucceeded;
+  payload: SuccessPayload;
+};
+
+export type OperationErroredAction = {
+  type: ActionTypes.OperationErrored;
+  payload: ErrorPayload;
+};
```

Comment:
I like that the action types are defined explicitly here! i tend to do this [indirectly with actions defined as functions](https://github.com/graphql/graphiql/pull/1463/files#diff-89da5d5b80d79f61b86bc31ea55c5555R17-R22) but this reads nicer

Reply:
haha i was thinking I liked the way you did it because you don't have to write them all out and it can be a bit more concise, but I guess it's also helpful to have them explicitly defined so if you go inspect the the definition you actually show up at a type instead of a function or something